new article on optimizing memory for bulk uploads (#1460)
qdrant/landing_page: qdrant-landing/content/articles/indexing-optimization.md

description: "Efficient memory management is key when handling large-scale vector data. Learn how to optimize memory consumption during bulk uploads in Qdrant and keep your deployments performant under heavy load."  
preview_dir: /articles_data/indexing-optimization/preview-4
social_preview_image: /articles_data/indexing-optimization/social-preview.png  
weight: -155
author: Sabrina Aquino  
date: 2025-02-12T00:00:00.000Z  
category: vector-search-manuals
---

# Optimizing Memory Consumption During Bulk Uploads

Efficient memory management is a constant challenge when you’re dealing with **large-scale vector data**. In high-volume ingestion scenarios, even seemingly minor configuration choices can significantly impact stability and performance.

Let’s take a look at the best practices and recommendations to help you optimize memory usage during bulk uploads in Qdrant. We'll cover scenarios with both **dense** and **sparse** vectors, helping your deployments remain performant even under high load and avoiding out-of-memory errors.

---

### Indexing for dense vs. sparse vectors

**Dense vectors**  

> Qdrant employs an **HNSW-based index** for fast similarity search on dense vectors. By default, HNSW is built or updated once the number of **unindexed** vectors in a segment exceeds a set `indexing_threshold`. Although it delivers excellent query speed, building or updating the HNSW graph can be **memory-intensive** if it occurs frequently or across many small segments.

Reviewer: It's resource intensive in general. A user can expect 100% CPU consumption especially on smaller clusters. Is that something we want to cover in this article as well, or is it purely focused on memory usage?

Author: Good point. This one focuses more on memory usage, but we could mention that it's resource-intensive overall, since that also ties into memory.

**Sparse vectors**

Sparse vectors use an **inverted index**. This index is updated at the **time of upsertion**, meaning you cannot disable or postpone it for sparse vectors. In most cases, its overhead is smaller than that of building an HNSW graph, but you should still be aware that each upsert triggers a sparse index update.

### Disabling vs. deferring dense indexing

**`indexing_threshold=0`** 
 
Disables HNSW index creation for dense vectors. Qdrant will not build the HNSW graph for those vectors, letting you upload large volumes of data without incurring the memory cost of index creation.

**`indexing_threshold>0`**  

> A positive threshold tells Qdrant how many unindexed dense vectors can accumulate in a segment before building the HNSW graph. Small thresholds (e.g., 100) mean more frequent indexing with less data each time, which can still be costly if many segments exist. Larger thresholds (e.g., 10000) delay indexing to batch more vectors at once, potentially using more RAM at the moment of index build, but fewer builds overall. 

Reviewer: It's probably good to mention the unit is kilobytes, as it's a bit arbitrary otherwise.

```suggestion
A positive threshold tells Qdrant how many kilobytes unindexed dense vectors can accumulate in a segment before building the HNSW graph. Small thresholds (e.g., 100 KB) mean more frequent indexing with less data each time, which can still be costly if many segments exist. Larger thresholds (e.g., 10000 KB) delay indexing to batch more vectors at once, potentially using more RAM at the moment of index build, but fewer builds overall. 

The following operation can be used to [update](https://qdrant.tech/documentation/concepts/collections/#update-collection-parameters) the indexing threshold in your existing collection:
```

Author: Thanks!

```json
PATCH /collections/your_collection
{
  "optimizer_config": {
    "indexing_threshold": 10000
  }
}
```

<aside role="status">Sparse vectors are always indexed on upsert, regardless of the threshold.</aside>

---

### The `"m"` parameter

For dense vectors, the `m` parameter defines how many edges each node in the HNSW graph can have. Setting `"m": 0` effectively **disables the HNSW graph**, so no dense vector index will be built, no matter the `indexing_threshold`. This can be helpful during massive ingestion if you don’t need immediate searchability.

---

## On-Disk Indexing in Qdrant

> By default, Qdrant keeps **vectors and indexes** in **memory** to ensure low-latency queries. However, in large-scale or memory-constrained scenarios, you can configure some or all of those indexes to be stored on-disk. This helps reduce RAM usage at the cost of potential increases in query latency, particularly for cold reads.

Reviewer: We can put a few different things on disk:
- vectors itself
- payload data
- payload index
- HNSW index

Reading the bullet list below it seems to only cover payload indices, while the title makes me suggest it also covers HNSW.

Since the title of the article is about memory usage for bulk upload, maybe it's worth to mention all of the above since all of them affect memory usage.

In that case I'd reword 'When to use on-disk indexing' to just 'When to use on-disk'.

Author: Nice catch, this can definitely be more clear.

**When to use on-disk indexing**:
- You have **very large** or **rarely used** payload indexes, and freeing up RAM is worth potential I/O overhead.  
- Your dataset doesn’t fit comfortably in available memory.  
- You can tolerate slower queries if it ensures the system remains stable under heavy loads.

---

## Memmap storage and segmentation

> Qdrant uses **memory-mapped files** (segments) to store data on-disk. Rather than loading all vectors into RAM, Qdrant maps each segment into its address space, paging data in and out on demand. This helps keep the active RAM footprint lower, but each segment still incurs overhead (metadata, page table entries, etc.).

Reviewer: ```suggestion
Qdrant uses **memory-mapped files** (segments) to store data on-disk. Rather than loading all vectors into RAM, Qdrant maps each segment into its address space, paging data in and out on demand. This helps keep the active RAM footprint lower, because data can be paged out if memory pressure is high. But each segment still incurs overhead (metadata, page table entries, etc.).
```

During **high-volume ingestion**, you can accumulate dozens of small segments. Qdrant’s **optimizer** can later merge these into fewer, larger segments, reducing per-segment overhead and lowering total memory usage.

When you create a collection with `"on_disk": true`, Qdrant will store newly inserted vectors in memmap storage from the start. For example:

```json
PATCH /collections/your_collection
{
    "vectors": {
      "on_disk": true
    }
}
```

> This approach immediately places all incoming vectors on disk. If you want both dense and sparse vectors to be stored on disk, you need to enable `on_disk` for each type separately.

Reviewer: ```suggestion
This approach immediately places all incoming vectors on disk, which can be very efficient in case of bulk ingestion. If you want both dense and sparse vectors to be stored on disk, you need to enable `on_disk` for each type separately.
```

> For dense vectors, set `on_disk: true` inside `hnsw_config`.

Reviewer: This is not strictly true, and maybe a bit confusing.

For both dense and sparse vectors you can set `on_disk: true` in the vector configuration itself, that will put the vector data on disk.

But, you can also set `on_disk: true` for the HNSW and sparse vector indices. They are separate things.

I'd recommend to adjust this sentence a bit to clarify this difference. In the examples below we can probably show setting both `on_disk` flags.

Author: Yeah that got confusing, I'll adjust to cover all use cases

```json
PATCH /collections/your_collection
{
    "hnsw_config": {
        "on_disk": true
    }
}
```

For sparse vectors, configure `on_disk` inside the `index` section of each sparse vector field.

```json
PATCH /collections/your_collection
{
    "sparse_vectors": {
        "text": {
            "index": {
                "on_disk": false
            }
        }
    }
}
```

---

## Best practices for high-volume vector ingestion

Bulk ingestion can lead to high memory consumption and even out-of-memory (OOM) errors. **If you’re experiencing out-of-memory errors with your current setup**, scaling up temporarily (increasing available RAM) will provide a buffer while you adjust Qdrant’s configuration for more a efficient data ingestion. 

> The key here is to control indexing overhead. Let’s walk through the best practices for high-volume vector ingestion in a constrained-memory environment.

Reviewer: Not strictly related to memory, but I think it's a good thing to mention and to understand in general. I'm not sure where a sentence like this fits best, and it probably needs a bit of rewording:

> The Qdrant optimizers always try to structure your data in a way that is efficient for searching. If you're bulk uploading a lot of data, Qdrant will constantly be moving all your data around to achieve the goal of efficient search even while your upload is still going on. This adds a lot of unnecessary data movement and overhead. Instead we recommend to upload all your data first, and only let the optimizers do their job after in a single go once all the data is there.

Author: I've added this as an additional point!

> 

Reviewer: The most significant change to help with memory is probably to put vector data on disk right away, with the `on_disk: true` flag as discussed above. I'd recommend to put that in this list as first item.

One or two years ago, when we were not able to put vectors on disk yet, vector data itself definitely was the biggest problem. Vector data is huge, and your were forced to keep it in memory before optimization. Optimizers simply weren't fast enough to process all the data and move it to disk.

Directly putting vectors on disk solves this problem. As far as I know, the memory overhead of HNSW indexing isn't that bad, unless you're talking about billion scale.

Author: done! yeah that's a major item. I thought just leaving it as a separate section would be enough but we should definitely have it on the best practices list as well!

**1. Disable HNSW for dense vectors (`m=0`)**

During an **initial bulk load**, you can **disable** dense indexing by setting `"m"=0`. This ensures Qdrant won’t build an HNSW graph for incoming vectors.

> Keep `indexing_threshold=10000` (or another large number) so that when you re-enable HNSW, you won’t trigger immediate, frequent index builds. This avoids memory spikes from HNSW building as data arrives, but leaves you the option to enable indexing later by setting `m` to a positive value.

Reviewer: I don't think it works this way.

I believe that setting either `indexing_threshold` or `m` back to it's original value will trigger indexing immediately on all your data.

It doesn't do it on all segments at once, but rather in a few jobs one after the other. The larger your CPU is, the more jobs it does in parallel. This is configurable using `max_indexing_threads`, but I'm not sure if it's worth to cover that here.

Ps, I personally prefer setting `indexing_threshold` instead as it's usually a bit easier to understand.

Author: interesting, I need to look deeper into it later on. thanks for the clarification!

```json
PATCH /collections/your_collection
{
  "hnsw_config": {
    "m": 0
  },
  "optimizer_config": {
    "indexing_threshold": 10000
  }
}
```

> <aside role="status">If your collection already contains a large number of vectors, changing these parameters will trigger a full index reconstruction, potentially causing downtime or performance degradation.</aside>

Reviewer: It should only cause downtime if nodes crash

```suggestion
<aside role="status">If your collection already contains a large number of vectors, changing these parameters will trigger a full index reconstruction, potentially causing slight performance degradation.</aside>
```

**2. Wait for indexation to clear up memory**

Allow Qdrant to finish any ongoing indexing before doing more operations. Large indexing jobs can keep memory usage high until they fully complete. Watch Qdrant logs or metrics to confirm when indexing finishes. Once that happens, memory consumption should drop as intermediate data structures are freed.

**3. Re-enable HNSW post-ingestion**

After the ingestion phase is over and memory usage has stabilized, re-enable HNSW for dense vectors by setting `m` back to a production value (commonly `16` or `32`):

```json
PATCH /collections/your_collection
{
  "hnsw_config": {
    "m": 16
  }
}
```

> **4. Enable quantization**

Reviewer: It's probably best to enable this before enabling indexing again, just to prevent having to do all optimizations a second time.

What would be even better is to enable indexing and quantization in the very same collection update call. But I understand that may be hard to explain in a nice and understandable way here.

Author: I'll add it as a tip!

> If you had planned to store all dense vectors on disk, be aware that searches can slow down drastically due to frequent disk I/O. A more balanced approach is **scalar quantization**: compress vectors (e.g., to `int8`) so they fit in RAM without occupying as much space as full floating-point values.

Reviewer: Nitpick: if a user has infinite memory, all data will just be kept in memory in which case I expect it to be fast.

I'll only become slow if a lot of memory pages are constantly swapped in and out because memory pressure is high / memory is limited.

```suggestion
If you had planned to store all dense vectors on disk, be aware that searches can slow down drastically due to frequent disk I/O while memory pressure is high. A more balanced approach is **scalar quantization**: compress vectors (e.g., to `int8`) so they fit in RAM without occupying as much space as full floating-point values.
```

Author: nice observation!

---

## Memmap storage and segmentation

Qdrant uses **memory-mapped files** (segments) to store data on-disk. Rather than loading all vectors into RAM, Qdrant maps each segment into its address space, paging data in and out on demand. This helps keep the active RAM footprint lower, but each segment still incurs overhead (metadata, page table entries, etc.).

During **high-volume ingestion**, you can accumulate dozens of small segments. Qdrant’s **optimizer** can later merge these into fewer, larger segments, reducing per-segment overhead and lowering total memory usage.

When you create a collection with `"on_disk": true`, Qdrant will store newly inserted vectors in memmap storage from the start. For example:

```json
PATCH /collections/your_collection
{
    "vectors": {
      "on_disk": true
    }
}
```

This approach immediately places all incoming vectors on disk. You can configure a threshold so that small segments are kept in RAM, but large segments automatically move to memmap storage once they exceed that threshold. This is controlled by the `memmap_threshold` in the collection’s `optimizers_config`.

```json
PATCH /collections/your_collection
{
    "optimizers_config": {

>         "memmap_threshold": 20000

Reviewer 2: `memmap_threshold` is going to be deprecated, there are no benefits in using it

Author: yeah, removed this part
